### PR TITLE
Refactor table size calculation and fix regressions

### PR DESCRIPTION
**Pull request by gokhangulbiz**
This pull request refactors the table size and table partitioning logic in the schema preparation process and stores the results in sqlite catalogs. 

It also disables caching feature for table sizes that needs to be implemented as a seperate PR as we discussed.

### CODE REVIEW

File: src/bin/pgcopydb/schema.c [hunk 1]
```diff
@@ -4759,6 +4626,99 @@ getTableArray(void *ctx, PGresult *result)
 }
 
 
+/*
+ * calculatePartKeyMinMaxValue calculates the min and max values for the
+ * candidate partition key of the given table.
+ */
+static bool
+calculatePartKeyMinMaxValue(PGSQL *pgsql, SourceTable *table)
```
Comment (dimitri) [inline src/bin/pgcopydb/schema.c:4634]: Naming: this does not look like a calculus, rather just fetching information from the source database. What about `getPartKeyMinMaxValue` instead?

File: src/bin/pgcopydb/copydb_schema.c [hunk 2]
```diff
@@ -751,6 +748,14 @@ copydb_prepare_table_specs_hook(void *ctx, SourceTable *source)
...
 			return false;
 		}
+
+		/* fetch the relpages for the table after ANALYZE */
+		if (!schema_get_relpages(context->pgsql, source))
+		{
+			log_error("Failed to fetch table %s relpages",
+					  source->qname);
+			return false;
+		}
```
[the remark this answers is not in the document]
Comment (gokhangulbiz) [reply src/bin/pgcopydb/copydb_schema.c]: [52a0b01](https://github.com/dimitri/pgcopydb/pull/684/commits/52a0b0113c63e45890633941f68860be6901df2c)

File: src/bin/pgcopydb/schema.c [hunk 3]
```diff
@@ -3377,122 +3216,65 @@ schema_list_partitions(PGSQL *pgsql,
 		return true;
 	}
 
-	PQExpBuffer sql = createPQExpBuffer();
+	/* if we have a partKey and it's not "ctid", calculate key bounds  */
```
[the remark this answers is not in the document]
Comment (gokhangulbiz) [reply src/bin/pgcopydb/schema.c:3229]: [9cb18a4](https://github.com/dimitri/pgcopydb/pull/684/commits/9cb18a4bfbc812f9b5052ac034335b0becb3c00d)

File: src/bin/pgcopydb/schema.c [hunk 4]
```diff
@@ -3377,122 +3216,65 @@ schema_list_partitions(PGSQL *pgsql,
...
+		min = 0;
+		max = table->relpages;
 	}
 
-	if (PQExpBufferBroken(sql))
+	int64_t partsCount = ceil((double) table->bytes / (double) partSize);
+	int64_t range = ceil((double) (max - min + 1) / (double) table->bytes *
+						 (double) partSize);
+
+	for (int64_t i = 0; i < partsCount; i++)
```
[the remark this answers is not in the document]
Comment (gokhangulbiz) [reply src/bin/pgcopydb/schema.c:3277]: [78fc1a6](https://github.com/dimitri/pgcopydb/pull/684/commits/78fc1a6d91f8ef0f5dfd63d0e0857c84dd9581af)

File: src/bin/pgcopydb/schema.c [hunk 5]
```diff
@@ -4715,6 +4497,90 @@ getCollationList(void *ctx, PGresult *result)
 }
 
 
+static void
+getTableSizeArray(void *ctx, PGresult *result)
```
[the remark this answers is not in the document]
Comment (gokhangulbiz) [reply src/bin/pgcopydb/schema.c:4539]: [0cbda55](https://github.com/dimitri/pgcopydb/pull/684/commits/0cbda55ef583ba29f0747eb9f25e93e5c5eeaeb1)

File: src/bin/pgcopydb/schema.c [hunk 6]
```diff
@@ -4715,6 +4497,90 @@ getCollationList(void *ctx, PGresult *result)
...
+			}
+		}
+	}
+
+	context->parsedOk = parsedOk;
+}
+
+
+static bool
+parseCurrentSourceTableSize(PGresult *result, int rowNumber, SourceTableSize *tableSize)
```
[the remark this answers is not in the document]
Comment (gokhangulbiz) [reply src/bin/pgcopydb/schema.c:4583]: [8ed4926](https://github.com/dimitri/pgcopydb/pull/684/commits/8ed4926c6b02882cb74464dae8f199932d95ec39)

File: src/bin/pgcopydb/schema.c (hunk 1 above)
Comment (gokhangulbiz) [reply src/bin/pgcopydb/schema.c]: [32b403d](https://github.com/dimitri/pgcopydb/pull/684/commits/32b403d8b537a81134e65583f6821e733997efbc)

File: src/bin/pgcopydb/catalog.c [hunk 7]
```diff
@@ -1948,6 +1962,68 @@ catalog_add_s_table_chksum(DatabaseCatalog *catalog,
 }
 
 
+bool
+catalog_add_s_table_size(DatabaseCatalog *catalog,
+						 SourceTableSize *tableSize)
```
[the remark this answers is not in the document]
Comment (gokhangulbiz) [reply src/bin/pgcopydb/catalog.c:1970]: [bad06f5](https://github.com/dimitri/pgcopydb/pull/684/commits/bad06f556c3a5e0e02a1db3609e7bbfaa1268919)

File: src/bin/pgcopydb/schema.c [hunk 8]
```diff
@@ -4759,6 +4626,99 @@ getTableArray(void *ctx, PGresult *result)
...
+	{
+		(void) destroyPQExpBuffer(sql);
+		log_error("Failed to parse SQL query to get partition key range");
+		return false;
+	}
+
+	(void) destroyPQExpBuffer(sql);
+
+	table->partmax = (partKeyMinMaxValueContext.max);
+	table->partmin = (partKeyMinMaxValueContext.min);
```
[the remark this answers is not in the document]
Comment (gokhangulbiz) [reply src/bin/pgcopydb/schema.c:4710]: I believe we already have this information (including partitions and its ranges) in s_table_parts. Let me know if you want me to duplicate this data in s_table.

File: src/bin/pgcopydb/schema.c [hunk 9]
```diff
@@ -3377,122 +3226,90 @@ schema_list_partitions(PGSQL *pgsql,
...
-	(void) destroyPQExpBuffer(sql);
+		parts->count = parts->max - parts->min + 1;
 
-	if (!parseContext.parsedOk)
-	{
-		log_error("Failed to list table COPY partition list");
-		return false;
+		log_info("Partition %s#%d: %lld - %lld (%lld)", table->qname, parts->partNumber,
+				 (long long) parts->min, (long long) parts->max, (long
+																  long) parts->count);
```
Comment (dimitri) [inline src/bin/pgcopydb/schema.c:3297]: That should be `log_debug`. Maybe could be raised to `log_notice`, but I'm not even sure it's that important as a log message.

File: src/bin/pgcopydb/schema.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (dimitri) [reply src/bin/pgcopydb/schema.c:4539]: Nitpicking: function comments should start with the function name.

File: src/bin/pgcopydb/schema.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (dimitri) [reply src/bin/pgcopydb/schema.c:4583]: Nitpicking: function comments should start with the function name.

File: src/bin/pgcopydb/schema.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (dimitri) [reply src/bin/pgcopydb/schema.c:4710]: Oh, right!

**Review by dimitri — CHANGES_REQUESTED** [review]
Very small changes required now, I missed a `log_info` that's too high in logging priority, should be decreased to `log_debug`... and that's about all there is to change now. Thanks again!

**Review by dimitri — APPROVED** [review]

**Pull request merged**